File: ass4.py
def generate_intermediate_code(assembly_code):
    intermediate_code = []
    symbol_table = {}
    data_allocation = {}
    current_address = 0
    data_offset = 0

    # First Pass: Process instructions and build symbol table
    lines = assembly_code.strip().split('\n')
    for line in lines:
        tokens = line.split()
        if len(tokens) == 0:
            continue

        mnemonic = tokens[0]

        if mnemonic == 'START':
            current_address = int(tokens[1])
        elif mnemonic == 'END':
            break
        elif mnemonic == 'DS':
            var_name = tokens[1]
            var_size = int(tokens[2])
            data_allocation[var_name] = data_offset
            data_offset += var_size
        elif mnemonic.endswith(':'):
            # Label definition
            label = mnemonic[:-1]
            symbol_table[label] = current_address
        else:
            # Instruction with operands
            op_code = mnemonic
            operands = ', '.join(tokens[1:])
            intermediate_code.append(f"{current_address}: {op_code} {operands}")
            current_address += 1

    # Second Pass: Resolve symbolic addresses and update data allocation
    for i, line in enumerate(intermediate_code):
        parts = line.split()
        if len(parts) > 1:
            op_code = parts[1]
            operands = ' '.join(parts[2:])
            for token in operands.split(', '):
                if token in symbol_table:
                    operands = operands.replace(token, str(symbol_table[token]))
            intermediate_code[i] = f"{parts[0]}: {op_code} {operands}"

    # Add data allocation information to the intermediate code
    for var_name, offset in data_allocation.items():
        intermediate_code.append(f"{current_address}: DS {var_name}, {offset}")
        current_address += 1

    return intermediate_code

# Example Assembly Language Code
assembly_code = """
START 100
READ A
READ B
MOVER AREG, A
SUB AREG, B
STOP
A DS 1
B DS 1
END
"""

# Generate intermediate code from assembly code
intermediate_code = generate_intermediate_code(assembly_code)

# Print intermediate code
for line in intermediate_code:
    print(line)

File: ass4.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>

using namespace std;

// Function to generate Intermediate code from assembly code
vector<string> generateIntermediateCode(const string& assemblyCode) {
    vector<string> intermediateCode;
    unordered_map<string, int> symbolTable;
    unordered_map<string, int> dataAllocation;
    int currentAddress = 0;
    int dataOffset = 0;

    // First Pass: Process instructions and build symbol table
    stringstream ss(assemblyCode);
    string line;
    while (getline(ss, line, '\n')) {
        if (line.empty()) {
            continue;
        }

        istringstream lineStream(line);
        string token;
        lineStream >> token;

        if (token == "START") {
            lineStream >> currentAddress;
        } else if (token == "END") {
            break;
        } else if (token == "DS") {
            string varName;
            int varSize;
            lineStream >> varName >> varSize;
            dataAllocation[varName] = dataOffset;
            dataOffset += varSize;
        } else if (token.back() == ':') {
            // Label definition
            string label = token.substr(0, token.length() - 1);
            symbolTable[label] = currentAddress;
        } else {
            // Instruction with operands
            string opcode = token;
            string operands;
            getline(lineStream >> ws, operands);
            operands.erase(remove(operands.begin(), operands.end(), ','), operands.end());

            // Replace labels with addresses in operands
            stringstream operandStream(operands);
            string operand;
            string updatedOperands;
            while (getline(operandStream, operand, ' ')) {
                if (symbolTable.find(operand) != symbolTable.end()) {
                    operand = to_string(symbolTable[operand]);
                }
                updatedOperands += operand + " ";
            }

            intermediateCode.push_back(to_string(currentAddress) + ": " + opcode + " " + updatedOperands);
            currentAddress++;
        }
    }

    // Second Pass: Update intermediate code with data allocation information
    for (auto& data : dataAllocation) {
        intermediateCode.push_back(to_string(currentAddress) + ": DS " + data.first + ", " + to_string(data.second));
        currentAddress++;
    }

    return intermediateCode;
}

int main() {
    // Example Assembly Language Code
    string assemblyCode = R"(
START 100
READ A
READ B
MOVER AREG, A
SUB AREG, B
STOP
A DS 1
B DS 1
END
)";

    // Generate intermediate code from assembly code
    vector<string> intermediateCode = generateIntermediateCode(assemblyCode);

    // Print intermediate code
    for (const string& codeLine : intermediateCode) {
        cout << codeLine << endl;
    }

    return 0;
}
